idea/src/org/jetbrains/kotlin/idea/intentions/JoinDeclarationAndAssignmentIntention.kt

```diff
@@ -75,25 +77,34 @@ class JoinDeclarationAndAssignmentIntention : SelfTargetingRangeIntention<KtProp
     }
 
     override fun applyTo(element: KtProperty, editor: Editor?) {
-        val typeReference = element.typeReference ?: return
-
+        if (element.typeReference == null) return
         val assignment = findAssignment(element) ?: return
         val initializer = assignment.right ?: return
-        val newInitializer = element.setInitializer(initializer)!!
 
+        element.initializer = initializer
         if (element.hasModifier(KtTokens.LATEINIT_KEYWORD)) element.removeModifier(KtTokens.LATEINIT_KEYWORD)
 
-        val initializerBlock = assignment.parent.parent as? KtAnonymousInitializer
-        assignment.delete()
-        if (initializerBlock != null && (initializerBlock.body as? KtBlockExpression)?.isEmpty() == true) {
-            initializerBlock.delete()
+        val grandParent = assignment.parent.parent
+        val initializerBlock = grandParent as? KtAnonymousInitializer
+        val secondaryConstructor = grandParent as? KtSecondaryConstructor
+        val newProperty = if (!element.isLocal && (initializerBlock != null || secondaryConstructor != null)) {
+            assignment.delete()
+            if ((initializerBlock?.body as? KtBlockExpression)?.isEmpty() == true) initializerBlock.delete()
+            if ((secondaryConstructor?.bodyBlockExpression)?.isEmpty() == true) secondaryConstructor.delete()
```

Comment:
Careful please. Constructor can have delegation, in this case you may delete its block but not constructor itself.

Before
```
constructor(...): this(...) {
    x = 42
}
```

After

```
constructor(...): this(...)
```